Mathlib/Algebra/Group/Ideal.lean

```diff
@@ -0,0 +1,186 @@
+/-
+Copyright (c) 2025 Dexin Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dexin Zhang
+-/
+import Mathlib.Algebra.Order.Monoid.Canonical.Defs
+import Mathlib.GroupTheory.GroupAction.SubMulAction
+import Mathlib.Order.WellFoundedSet
+
+/-!
+# Semigroup ideals
+
+This file defines (left) semigroup ideals (also called monoid ideals sometimes), which are sets `s`
+in a semigroup such that `a * b ∈ s` whenever `b ∈ s`. Note that semigroup ideals are different from
+ring ideals, where a ring ideal must also be an additive submonoid of the ring.
+
+## Main definitions
+
+* `SemigroupIdeal`: a (left) semigroup ideal in a semigroup `M` is a set `s` such that `a * b ∈ s`
+  whenever `b ∈ s`.
+* `SemigroupIdeal.closure`: the semigroup ideal generated by a set `s`.
+* `SemigroupIdeal.FG`: a semigroup ideal is finitely generated if it is generated by a finite set.
+
+## Main results
+
+* In a canonically ordered and well-quasi-ordered monoid, any semigroup ideal is finitely generated,
+  and the semigroup ideals satisfy the ascending chain condition.
+
+## References
+
+* [Samuel Eilenberg and M. P. Schützenberger, *Rational Sets in Commutative Monoids*][eilenberg1969]
+-/
+
+open Set Pointwise
+
+/-- A (left) semigroup ideal in a semigroup `M` is a set `s` such that `a * b ∈ s` whenever
+`b ∈ s`. -/
+@[to_additive /-- A (left) additive semigroup ideal in an additive semigroup `M` is a set `s` such
+that `a + b ∈ s` whenever `b ∈ s`. -/]
+abbrev SemigroupIdeal (M : Type*) [Semigroup M] :=
+  SubMulAction M M
+
+namespace SemigroupIdeal
+
+variable {M : Type*}
+
+section Semigroup
+
+variable [Semigroup M]
+
+@[to_additive]
+protected theorem mul_mem (I : SemigroupIdeal M) (x : M) {y : M} : y ∈ I → x * y ∈ I :=
+  SubMulAction.smul_mem I x
+
+/-- The semigroup ideal generated by a set `s`, defined as `s ∪ Set.univ * s`. -/
+@[to_additive /-- The additive semigroup ideal generated by a set `s`, defined as
+`s ∪ Set.univ + s`. -/]
+def closure (s : Set M) : SemigroupIdeal M where
```

Comment:
Shouldn't you define this at the level of `SubMulAction` first?

Reply:
I mean that `SemigroupIdeal.closure` should be defined as an `abbrev` of `SubMulAction.closure`, which currently doesn't exist